consensus/view_change.go

```diff
@@ -90,20 +93,123 @@ func (pm *State) GetViewChangeDuraion() time.Duration {
+		return consensus.current.GetCurBlockViewID() + 1, time.Duration(int64(viewChangeDuration))
+	}
+	diff := int64(viewID + 1 - consensus.current.GetCurBlockViewID())
+	return viewID + 1, time.Duration(diff * diff * int64(viewChangeDuration))
+}
+
+// getNextViewID return the next view ID based on the timestamp
+// The next view ID is calculated based on the difference of validator's timestamp
+// and the block's timestamp. So that it can be deterministic to return the next view ID
+// only based on the blockchain block and the validator's current timestamp.
+// The next view ID is the single factor used to determine
+// the next leader, so it is mod the number of nodes per shard.
+// It returns the next viewID and duration of the view change
+// The view change duration is a fixed duration now to avoid stuck into offline nodes during
+// the view change.
+// viewID is only used as the fallback mechansim to determine the nextViewID
+func (consensus *Consensus) getNextViewID(viewID uint64) (uint64, time.Duration) {
+	// handle corner case at first
+	if consensus.ChainReader == nil {
+		return consensus.fallbackNextViewID(viewID)
+	}
+	curHeader := consensus.ChainReader.CurrentHeader()
+	if curHeader == nil {
+		return consensus.fallbackNextViewID(viewID)
+	}
+	blockTimestamp := curHeader.Time().Int64()
+	curTimestamp := time.Now().Unix()
+
+	// timestamp messed up in current validator node
+	if curTimestamp < blockTimestamp {
+		return consensus.fallbackNextViewID(viewID)
+	}
+	totalNode := consensus.Decider.ParticipantsCount()
+	// diff is at least 1, and it won't exceeded the totalNode
+	diff := uint64(((curTimestamp - blockTimestamp) / viewChangeTimeout) % int64(totalNode))
+	nextViewID := diff + consensus.current.GetCurBlockViewID()
+
+	consensus.getLogger().Info().
+		Int64("curTimestamp", curTimestamp).
+		Int64("blockTimestamp", blockTimestamp).
+		Uint64("nextViewID", nextViewID).
+		Uint64("curViewID", consensus.current.GetCurBlockViewID()).
+		Msg("[getNextViewID]")
+
+	// duration is always the fixed view change duration for synchronous view change
+	return nextViewID, viewChangeDuration
+}
+
+// getNextLeaderKey uniquely determine who is the leader for given viewID
+// It reads the current leader's pubkey based on the blockchain data and returns
+// the next leader based on the gap of the viewID of the view change and the last
+// know view id of the block.
+func (consensus *Consensus) getNextLeaderKey(viewID uint64) *bls.PublicKeyWrapper {
 	gap := 1
+
+	if viewID > consensus.GetCurBlockViewID() {
+		gap = int(viewID - consensus.GetCurBlockViewID())
+	}
+	var lastLeaderPubKey *bls.PublicKeyWrapper
+	var err error
+	epoch := big.NewInt(0)
+	if consensus.ChainReader == nil {
+		consensus.getLogger().Error().Msg("[getNextLeaderKey] ChainReader is nil. Use consensus.LeaderPubKey")
+		lastLeaderPubKey = consensus.LeaderPubKey
+	} else {
+		curHeader := consensus.ChainReader.CurrentHeader()
```

Comment:
If we can't get currentHeader, it's an error case which we should just return error since we don't know which epoch anymore. Epoch = 0 won't work.

Reply:
I think it may be the bootstrap phase, and view change happened.  It is not applicable to current mainnet though.

Reply:
how to calculate the right epoch then if the header is not available? that's why I use epoch 0 as the fallback value. in case of this error case happened.